src/plugins/discover/public/application/helpers/get_fields_to_show.ts

```diff
@@ -7,7 +7,18 @@
  */
 import { IndexPattern } from '../../../../data/common';
 
-export const getFieldsToShow = (
+const memoize = (func: (arg1: string[], arg2: IndexPattern, arg3: boolean) => string[]) => {
+  const results: Record<string, string[]> = {};
+  return (...args: [string[], IndexPattern, boolean]) => {
+    const argsKey = JSON.stringify(args);
```

Comment:
yes, this is much better 👍 
however sorry to be performance pedant, the stringify of those arguments would still be executed for each row, for our test index pattern this would add a few ms execution time, but I didn't test for larger index patterns, also they'd stay in memory. So I'd suggest we'd calculate getFieldsToShow on a higher level, passing it down. So it would just be executed once, no memoizing would be necessary.

Here it could be calculated and used instead of 

services.uiSettings.get(SHOW_MULTIFIELDS)

https://github.com/elastic/kibana/blob/77ec43737977df265ea5330ea33bc59c9fa48676/src/plugins/discover/public/application/components/discover_grid/discover_grid.tsx#L264-L270

same here, instead of `showMultiFields`, eg. pass down fieldsToShow

https://github.com/elastic/kibana/blob/225cfb9a8293341fba259bcaf22d8edf44185381/src/plugins/discover/public/application/apps/main/components/doc_table/doc_table_wrapper.tsx#L185-L200

Reply:
in `getRenderCellValueFn` we're reading the fields off from an individual row, which is a flattened hit. I don't know if it's a safe assumption to make that fields will be the same across all rows?

Reply:
<img width="1286" alt="Screenshot 2021-08-24 at 10 29 33" src="https://user-images.githubusercontent.com/1937956/130584118-4dbbf39c-06b8-454e-91b2-9229de65b9f2.png">

Reply:
No guarantee that the fields will be the same across rows, that's why we're formatting each row individually.

Reply:
So we'd still need some sort of a manual memoization in memory, like a map of all fields we've encountered so far and if they should be shown or not. We can move this then higher up and just update the map when each row renders its fields. Let me know what you think about this approach.

Reply:
Are there fields that are in the document but not provided by the index pattern fields? because I think it should behave like in classic table, use the fields of the index pattern to create the list of fields that should be displayed

Reply:
we agreed offline to move the code to a higher level in the execution, no memo required here